devolutions-gateway/src/http/controllers/kdc_proxy.rs

```diff
@@ -0,0 +1,160 @@
+        let address_to_resolve = format!(
+            "{}:{}",
+            kdc_proxy_config.kdc.host().unwrap().to_string(),
+            kdc_proxy_config.kdc.port().unwrap_or(DEFAULT_KDC_PORT)
+        );
+
+        let kdc_address = if let Some(address) = lookup_kdc(&address_to_resolve) {
+            address
+        } else {
+            slog_scope::error!("Unable to locate KDC server");
+            return Err(HttpErrorStatus::internal("Unable to locate KDC server"));
+        };
+
+        let mut kdc_reply_message = Vec::new();
+
+        if scheme == "tcp" {
+            let mut connection = TcpStream::connect(kdc_address).await.map_err(|e| {
+                slog_scope::error!("{:?}", e);
+                HttpErrorStatus::internal("Unable to connect to KDC server")
+            })?;
+
+            connection
+                .write_all(&kdc_proxy_message.kerb_message.0 .0)
+                .await
+                .map_err(|e| {
+                    slog_scope::error!("{:?}", e);
+                    HttpErrorStatus::internal("Unable to send the message to the KDC server")
+                })?;
+
+            read_kdc_reply_message(&mut connection, &mut kdc_reply_message).await?;
+        } else if scheme == "udp" {
+            let mut buff = vec![0; 1024];
+
+            let port = portpicker::pick_unused_port().ok_or_else(|| HttpErrorStatus::internal("No free ports"))?;
+            let udp_socket = UdpSocket::bind(format!("127.0.0.1:{}", port)).await.map_err(|e| {
+                slog_scope::error!("{:?}", e);
+                HttpErrorStatus::internal("Unable to send the message to the KDC server")
+            })?;
+
+            // first 4 bytes contains message length. we don't need it for UDP
+            udp_socket
+                .send_to(&kdc_proxy_message.kerb_message.0 .0[4..], kdc_address)
+                .await
+                .map_err(|e| {
+                    slog_scope::error!("{:?}", e);
+                    HttpErrorStatus::internal("Unable to send the message to the KDC server")
+                })?;
+
+            let n = udp_socket.recv(&mut buff).await.map_err(|e| {
+                slog_scope::error!("{:?}", e);
+                HttpErrorStatus::internal("Unable to read reply from the KDC server")
+            })?;
+
+            kdc_reply_message.extend_from_slice(&(n as u32).to_be_bytes());
+            kdc_reply_message.extend_from_slice(&buff[0..n]);
+        }
+
+        let kdc_proxy_reply_message = KdcProxyMessage::from_raw_kerb_message(&kdc_reply_message).map_err(|e| {
+            slog_scope::error!("{:?}", e);
+            HttpErrorStatus::internal("Cannot create kdc proxy massage")
+        })?;
+
+        Ok(Builder::new()
+            .body(kdc_proxy_reply_message.to_vec().unwrap())
+            .status(200))
+    }
+}
+
+async fn read_kdc_reply_message(
+    connection: &mut TcpStream,
+    kdc_reply_message: &mut Vec<u8>,
+) -> Result<(), HttpErrorStatus> {
```

Comment:
General comment for this function: I don't think you need an intermediate buffer. Am I wrong?

Reply:
We need to read a reply from the `KDC` server, then wrap the reply into `KdcProxyMessage`, and send it back to the client. So a buffer is needed. Do you have different a approach?

Reply:
Notice in my suggestion there is no risk to read past the size of the message (I think we may eat to much bytes in your version, but I don't know if it's an actual problem)

Reply:
Done. Thank you.